pages/builds/environment_variables.md.erb

```diff
@@ -4,51 +4,228 @@ When the agent invokes your build scripts it passes in a set of standard Buildki
 
 * Pipeline settings — for values that are not secret.
 * [Build pipeline configuration](/docs/pipelines/command-step) — for values that are not secret.
 * An `environment` or `pre-command` [agent hook](/docs/agent/hooks) — for values that are secret or agent-specific.
+
+## Writable variables for specific purposes:
+
+Using S3 (see our [S3 docs](/docs/agent/cli-artifact#using-your-own-s3-bucket) for more information):
+BUILDKITE_ARTIFACT_UPLOAD_DESTINATION
+BUILDKITE_S3_DEFAULT_REGION
+BUILDKITE_S3_ACCESS_KEY_ID
+BUILDKITE_S3_SECRET_ACCESS_KEY
+BUILDKITE_S3_ACL
+
+Using Google Cloud (see our [google cloud docs](/docs/agent/cli-artifact#using-your-own-private-google-cloud-bucket) for more information):
+BUILDKITE_ARTIFACT_UPLOAD_DESTINATION
+BUILDKITE_GS_* <Needs more info>
+
+## How environment variables do the thing
+
+- you can set env vars in lots of different places, which ones take precedence can get a little confusing
+- there are many different levels at which environment variables are merged together
+- this example demonstrates the order in which env vars are combined, as if you had set variables in every possible place
+
+- when a job runs on an agent, the first step when putting together the environment happens in the job itself:
+
+### Job environment
+
+- a base environment is created for every job
+- this is environment that Buildkite returns if you look at a job using the REST API
+- this env made up of at least one, and up to four different sets of variables combined together: 
+
+standard: the set of variables provided by Buildkite to every job (https://github.com/buildkite/buildkite/blob/master/app/models/job/environment.rb#L32)
+optional: variables set on the build by you in when creating a new build in the UI or via the REST API
+optional: variables set on a step in the UI or in a pipeline.yml file by you
+optional: variables set on a pipeline in the Pipeline Settings page by you
+
+- these are merged together in the following way:
+(maybe put an image or example code here)
+bk env (most important, not overridden by the following bits) > build > step > project (least important, can be overridden by everything else)
+
+- more details on setting vars in a pipeline.yml:
+- there are two places in a pipeline.yml file that you can set env vars:
+  - in the `env` attribute of command or trigger steps
+  - at the top of the yaml file, before you define your pipeline's steps (this will set the defined variable on each of the command steps in the pipeline, and is equivalent to setting the `env` attribute on every step)
+
+- more details on setting vars in a Trigger step:
+  - env vars set in a trigger step end up as build-level env on the newly created build
+  - NOTE: the 'all steps env' from the top of a pipeline.yaml isn't applied to trigger steps, the only env that applies to trigger steps is what you choose to set in that trigger step's `env` attribute (which ends up as build env in the new build)
```

Comment:
Maybe worth adding that you can copy env to the trigger env using env interpolation from agent 3:

```yaml
env:
- FOO=bar
steps:
- trigger: some-pipeline
  build:
    env:
    - FOO=$FOO
```

and that trigger env becomes "build env" in the triggered build